Created indexes.sql + documentation

curious-bryan opened the pull request:
This commit is in response to Issues https://github.com/jacobm001/dbt-mssql/issues/12 and  https://github.com/jacobm001/dbt-mssql/issues/13.

### NOTES 

SQL file for macros for working with indexes on MS SQL Server.

### Macro drop_all_indexes_on_table:

This code was adapted from Stackoverflow answer at https://stackoverflow.com/q/1344401/10415173.

**Code follows these steps**

1. Drop all XML indexes (if any)
2. Drop all spatial indexes (if any)
3. Drop all foreign keys pointing to table (if any)
4. Drop all primary key constraints (if any)
5. Drop all remaining indexes

**Uses these system objects on MSSQL db:** 
- sys.indexes
- sys.tables
- sys.foreign_keys
- sp_executesql

### Macro create_nonclustered_index:

Creates nonclustered index with (or without) include clause.
This macro may need updating in the future for quoting columns.
Not sure how it will do for columns names with spaces but quoted in quotes or brackets.

curious-bryan committed:
Merge pull request #1 from jacobm001/master

Merge latest updates from 1.0.6

curious-bryan committed:
Create drop_all_indexes_on_table.sql

This code was adapted from Stackoverflow answer at https://stackoverflow.com/q/1344401/10415173.

Code follows these steps

1. Drop all XML indexes (if any)
2. Drop all spatial indexes (if any)
3. Drop all foreign keys pointing to table (if any)
4. Drop all primary key constraints (if any)
5. Drop all remaining indexes

Uses these system objects on MSSQL db:
- sys.indexes
- sys.tables
- sys.foreign_keys
- sp_executesql

curious-bryan committed:
Delete drop_all_indexes_on_table.sql

Delete this file and move all into one file called indexes.sql

curious-bryan committed:
Create indexes.sql

SQL file to contain macros for working with indexes on MS SQL Server.

Macro drop_all_indexes_on_table:

This code was adapted from Stackoverflow answer at https://stackoverflow.com/q/1344401/10415173.

Code follows these steps

1. Drop all XML indexes (if any)
2. Drop all spatial indexes (if any)
3. Drop all foreign keys pointing to table (if any)
4. Drop all primary key constraints (if any)
5. Drop all remaining indexes

Uses these system objects on MSSQL db:
- sys.indexes
- sys.tables
- sys.foreign_keys
- sp_executesql

Macro create_nonclustered_index:

Creates nonclustered index with (or without) include clause.
This macro may need updating in the future for quoting columns.
Not sure how it will do for columns names with spaces but quoted in quotes or brackets.

curious-bryan committed:
Fix spacing in indexes.sql

Changes tab spacing to 4-space spacing.

curious-bryan committed:
Add documentation about index macros

jacobm001 commented on dbt/include/mssql/macros/indexes.sql:
```diff
@@ -0,0 +1,76 @@
+{% macro drop_all_indexes_on_table() %}
```
This is a great start, but is a giant function. While it may make sense to have a "drop all" function, this is large and is actually doing multiple things (dropping different kinds of indexes). I'd like to see each of these steps moved into their own functions, all of which can be called from the "drop all" function.

curious-bryan replied:
Great feedback!  Will do.  May be a few days before I get to it.

jacobm001 commented on dbt/include/mssql/macros/indexes.sql:
```diff
@@ -0,0 +1,76 @@
+{% macro drop_all_indexes_on_table() %}
```
This adds nothing to the debug log. Each step should produce a note for the logger so that users can see what's happening.

curious-bryan replied:
So true!  Didn't even think about that.  Will do.  May be a few days before I get to it.

jacobm001 reviewed (changes requested):
This is a great start! I think there are a few things that need to be cleaned up, but it's most of the way there.

curious-bryan committed:
Break macro into smaller macros and add logging

- Broke drop_all_indexes_on_table() into smaller macros where drop_all_indexes_on_table calls the smaller macros within it.
- Added simple logging to log file for different steps in process

curious-bryan commented:
Okay, @jacobm001 , I had a second to go ahead and make the changes in a new commit.  I added logging messages to each of the smaller macros and the create_nonclustered_index() macro.  Let me know what you think.  I don't mind changing the logging messages.

curious-bryan commented:
@jacobm001 , also, I am not as expert as T-SQL, so let me know if I should include the `drop_xml_indexes()` and `drop_spatial_indexes()` macros in the `drop_fk_constraints()` macro.  I have never created xml or spatial indexes.  But I was thinking they were only created against primary key indexes not foreign key constraints.

If I need to add the to `drop_fk_constraints()` macro, let me know and I will make those changes.

curious-bryan committed:
Added clustered_index macro

Added a simple macro for creating clustered indexes

curious-bryan committed:
Added documentation on clustered index macro

curious-bryan committed:
Add bullets to macro syntax section on indexes

jacobm001 commented on dbt/include/mssql/macros/indexes.sql:
```diff
@@ -0,0 +1,132 @@
+{% macro create_clustered_index(column) -%}
```
Why are we only allowing a single column to be added to the clustered index? These don't typically get very deep, but it's certainly not uncommon to use 2-3 columns for this.

Example:
A student transcript would likely be clustered by `["term", "student_id"]`

curious-bryan replied:
@jacobm001 , good point.  New commit made updating for more than one column.

jacobm001 commented on dbt/include/mssql/macros/indexes.sql:
```diff
@@ -0,0 +1,132 @@
+
+
+{% macro drop_pk_constraints() %}
+{# Altered from https://stackoverflow.com/q/1344401/10415173 #}
```
+1 for citing your sources!

jacobm001 reviewed (changes requested):
This is pretty close. Good work!

curious-bryan committed:
Updated clustered index

Updated clustered index to include more than one column

jacobm001 commented:
These changes look good! I'm merging into a new branch, so I can run the test harness against the changes and do a live test. Assuming it all works I'll have the new version out by Friday :)
